### DISCUSSION TITLE
Permissions denied (public key)

### DISCUSSION THREAD

**Discussion by derpyder** [gtxaspec/wz_mini_hacks#96] [Q&A] [answered]
As the title suggests, I'm stuck trying to SSH into the camera after setting up pub and private keys.

This command:

ssh -i /config/.ssh/privatekey.ppk root@[myipaddress] 

bounces back the denial. Any ideas what could be the issue with my generated keys? (did the puttygen method and saved the private key as openSSH)

Thanks in advance!

**Comment by sideup66** [comment]
ssh commandline wont read a ppk file, try saving the key as a .pem. I believe putty doesnt do any weird encoding so a simple rename or move should work :).

**Comment by eddieyee** [comment]
I get access denied as well.  I had generated my public key way before (done this many times on other mac/linux systems).   I tried copying  in the public key a few different ways and get "Permission denied (publickey)".   

In the simplest form, even doing a cat ~/.ssh/id_rsa.pub >> /Volumes/WYZE/wz_mini/etc/ssh/authorized_keys    gives the same error when trying to connect (from macos Monterey).   I also tried doing this from a Linux VM, different ssh pub/private keys and still had the same problem.

Interestingly, after trying to connect, the file system in the ssh folder will create a dropbear_ed25519_host_key.   (identifier in the middle changes depending on the host trying to connect).

**Reply by virmaior** [reply]
it took a bit to get it working on osx monterrey.

1. are doing sudo ssh or just ssh? (at least for me it only works with ssh)
2. did you start the ssh agent?
3. try ssh -v 192.168.9.101 (or whatever cameras IP address is)

Here's what a successful ssh looks like:
```
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: myusername@MacBook-Air.local RSA SHA256:2CJ9Phciq6AIyq9098J8MDREpdQ0e1dnoBzxXrOCuLc agent
debug1: Will attempt key: /var/root/.ssh/id_rsa 
debug1: Will attempt key: /var/root/.ssh/id_dsa 
debug1: Will attempt key: /var/root/.ssh/id_ecdsa 
debug1: Will attempt key: /var/root/.ssh/id_ecdsa_sk 
debug1: Will attempt key: /var/root/.ssh/id_ed25519 
debug1: Will attempt key: /var/root/.ssh/id_ed25519_sk 
debug1: Will attempt key: /var/root/.ssh/id_xmss 
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com,rsa-sha2-256,ssh-rsa,ssh-dss>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Offering public key: myusername@MacBook-Air.local RSA SHA256:XXX agent
debug1: Server accepts key: myusername@MacBook-Air.local RSA SHA256:XXXX agent
debug1: Authentication succeeded (publickey).
Authenticated to 192.168.9.101 ([192.168.9.101]:22).
debug1: channel 0: new [client-session]
debug1: Entering interactive session.
debug1: pledge: filesystem full
debug1: Sending environment.
debug1: channel 0: setting env LC_CTYPE = "UTF-8"
```

**Reply by eddieyee** [reply]
I was doing a normal ssh (non sudo).   Ssh agent is confirmed up and running, just not authenticating.  I generated keys by running ssh-keygen, copied the contents of id_pub.rsa to the sdcard /wz_mini/etc/ssh/authorized_keys.

At this point, the only thing I can think of is something maybe to do with the  hostname.  I'm using the ISP provided router, and it tacks on .attlocal.net to all my local hosts, and does some dnsmasq and local lookups on the router.   It does the same on a linux vm.

Here's the debug output.

debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /Users/eddie/.ssh/id_rsa RSA SHA256:OBpQqWPK+uAesEBhwgZ0mbw0o5cJcFE+cf3D1bYsz0I
debug1: Will attempt key: /Users/eddie/.ssh/id_dsa 
debug1: Will attempt key: /Users/eddie/.ssh/id_ecdsa 
debug1: Will attempt key: /Users/eddie/.ssh/id_ecdsa_sk 
debug1: Will attempt key: /Users/eddie/.ssh/id_ed25519 
debug1: Will attempt key: /Users/eddie/.ssh/id_ed25519_sk 
debug1: Will attempt key: /Users/eddie/.ssh/id_xmss 
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com,rsa-sha2-256,ssh-rsa,ssh-dss>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Offering public key: /Users/eddie/.ssh/id_rsa RSA SHA256:OBpQqWPK+uAesEBhwgZ0mbw0o5cJcFE+cf3D1bYsz0I
debug1: Authentications that can continue: publickey
debug1: Trying private key: /Users/eddie/.ssh/id_dsa
debug1: Trying private key: /Users/eddie/.ssh/id_ecdsa
debug1: Trying private key: /Users/eddie/.ssh/id_ecdsa_sk
debug1: Trying private key: /Users/eddie/.ssh/id_ed25519
debug1: Trying private key: /Users/eddie/.ssh/id_ed25519_sk
debug1: Trying private key: /Users/eddie/.ssh/id_xmss
debug1: No more authentication methods to try.
eddie@192.168.1.129: Permission denied (publickey).


My authorized_keys last part after the key shows:
eddie@Eddies-Mac-mini.attlocal.net

Not sure if it has any bearing on this, but I'm guessing there may be.

**Comment by dudududodododedede** [comment]
I don't think the terminal ssh command supports .ppk, it needs to be converted (renaming is not enough)
https://tecadmin.net/convert-ppk-to-pem-using-command/

Also try running ssh command in verbose mode to see debug messages

**Reply by derpyder** [reply]
negative on pem perhaps? Excuse the screengrab. Homesite terminal isn't letting me use any hotkeys to paste.
![msedge_EBG95jaUV0](https://user-images.githubusercontent.com/23619252/174419032-9ef6e315-b8f0-4293-a92c-9e0a40c02974.png)

**Comment by eddieyee** [comment]
I am wondering if there's another problem since I'm using standard ssh-keygen on my Mac.   I also tested on a linux system and it didn't like the authorized key that i copied in from my id_rsa.pub file.

**Comment by gtxaspec** [comment]
questions:
which model camera?
have you tried reformatting your memory card? (dos partition table, fat32 partition)

**Comment by eddieyee** [comment]
I am on a Wyze V3.  Camera is a few days old and works fine.  I've tried reformatting twice (dos fat32).

**Reply by derpyder** [reply]
Happy (and sad) I am not alone.

**Reply by gtxaspec** [reply]
really strange, may be something with your operating system.  works well running linux, debian  & ubuntu that I have tested.

**Comment by dudududodododedede** [comment]
things to check: https://serverfault.com/questions/854208/ssh-suddenly-returning-invalid-format

**Comment by derpyder** [comment, marked as answer]
I was able to resolve the authentication.  

Instead of adding the public key info to the authenticatedkeys file, i replaced that file entirely after using "ssh-keygen -t ed25519" to spit out my new pair of keys. These worked right out of the gate.

I was then running into the RTSP stream only spitting out audio, and after ditching the beta RTSP firmware for the Wyze stock, this too was resolved, in case anyone missed that in the latest 6/19 readme.

**Reply by eddieyee** [reply]
I tried generating the ed25519 formatted keypair, and it still has the same problem for me.  My verbose log shows it presenting the public key, but still gets denied.   Tested from Macos Monterrey as well as an Ubuntu 20.04 VM (Anybody know how to look at the dropbear logs on the camera to see what it sees, or if there's a way to enable verbose logging on the ssh connection from the camera?

Here's an ssh -v output from my Ubuntu host:

debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /home/eddie/.ssh/id_rsa RSA SHA256:9a6qbXcurDEgsOPKzfQNVBmhBy5kzja2sfQI6co4i0c
debug1: Will attempt key: /home/eddie/.ssh/id_dsa 
debug1: Will attempt key: /home/eddie/.ssh/id_ecdsa 
debug1: Will attempt key: /home/eddie/.ssh/id_ecdsa_sk 
debug1: Will attempt key: /home/eddie/.ssh/id_ed25519 ED25519 SHA256:GE+dKc2KVzUCOxjzGhZ6GjxFP+sdheYRCP4iRKW3A4E
debug1: Will attempt key: /home/eddie/.ssh/id_ed25519_sk 
debug1: Will attempt key: /home/eddie/.ssh/id_xmss 
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com,rsa-sha2-256,ssh-rsa,ssh-dss>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Offering public key: /home/eddie/.ssh/id_rsa RSA SHA256:9a6qbXcurDEgsOPKzfQNVBmhBy5kzja2sfQI6co4i0c
debug1: Authentications that can continue: publickey
debug1: Trying private key: /home/eddie/.ssh/id_dsa
debug1: Trying private key: /home/eddie/.ssh/id_ecdsa
debug1: Trying private key: /home/eddie/.ssh/id_ecdsa_sk
debug1: Offering public key: /home/eddie/.ssh/id_ed25519 ED25519 SHA256:GE+dKc2KVzUCOxjzGhZ6GjxFP+sdheYRCP4iRKW3A4E
debug1: Authentications that can continue: publickey
debug1: Trying private key: /home/eddie/.ssh/id_ed25519_sk
debug1: Trying private key: /home/eddie/.ssh/id_xmss
debug1: No more authentication methods to try.
eddie@192.168.1.129: Permission denied (publickey).

**Reply by eddieyee** [reply]
I revisited it and found there was nothing wrong with my keypair.  The problem was ssh... If I "ssh root@IP" from my host, it connects fine.
